### DISCUSSION TITLE
Conscious Language

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#8672] [Ideas]
The terms "master/slave" and "whitelist/blacklist" has been identified as the most visible and problematic across the industry, and we should plan to remove these terms from Keycloak as a first step.

As a first step towards more inclusive language in Keycloak we have identified the following tasks:

- [x] Rename "master" branches in GitHub repositories to "main"
- [ ] [Rename "master" realm](https://github.com/keycloak/keycloak/issues/20066)
- [ ] Rename "blacklist" password policy to "denylist"
- [ ] Review/update the above terms from documentation

**Comment by stianst** [comment]
Renaming branches in GitHub should be relatively straightforward as GitHub provided extensive tooling to support this effort. See https://github.com/github/renaming for more details.

**Reply by stianst** [reply]
All default branches are now renamed to "main"

**Comment by stianst** [comment]
For "blacklist" password policy we would need to find a way to make this a non-breaking change. We could either have both a deprecated blacklist password policy, and a new "denylist" password policy, or perhaps better we rename and migrate existing realms, including also handling the rename in imports of realms.

**Reply by ZacharyACoon** [reply]
Just keep support for `blacklist`, add support for `denylist`, and change all display language to prefer `denylist`, mark `blacklist` as deprecated in documentation.  Done.

**Comment by stianst** [comment]
For renaming the master realm to something else, we could call it main, but perhaps there is a name that can better reflect that this is a "Keycloak admin" realm, and not a realm that is recommended to be used for general purposes.

Renaming the master realm will regardless be a breaking change, so we may want to only do this for new installations, with a manual process to allow renaming the realm for existing installations.

**Reply by jbman** [reply]
It would be good to have a **realm public name** used for authorization and admin URLs. As long as not specified, the public name equals the realm name. A new version of Keycloak should ship with realm name `main` and add the public name `master` to the main realm on migration.

Such a feature would also help to change a realm URL without breaking their IdenitityProvider integrations. We often support customers with realm creation and configuration and later someone wants to improve the URL and renames the realm without knowing that it breaks IdP OIDC clients. With the realm public name, the initial realm name could become unrename-able in the admin console and the public name can be managed instead.

**Reply by sre4ever** [reply]
Good point, we sometimes have that use case as well (renaming realms) but I don't like the idea of making the initial realm name immutable. Instead, maybe plan a feature like realm aliases (allowing several ones would make it easier to merge realms).
It would also be nice to provide a way to change the name of the administration realm through configuration.

**Reply by stianst** [reply]
Actually, what I've wanted for a long time is the ability to map arbitrary URLs to a realm. For example https://idp.acme.org could be mapped to the realm "acme-external", without needing the "/realms/acme-external" part in the URL at all. Same thing could be used to map "/realms/master" to the main realm.

**Reply by ZacharyACoon** [reply]
`admin` realm.   Or just allow the admin to use any realm as the default/admin realm.

**Comment by vmuzikar** [comment]
I believe this concerns also the "master" tag of the [operator quay repo](https://quay.io/repository/keycloak/keycloak-operator). The risk here is it is potentially a breaking change.
